src/libopensc/pkcs15.c

```diff
@@ -2378,22 +2400,43 @@ sc_pkcs15_read_file(struct sc_pkcs15_card *p15card, const struct sc_path *in_pat
 
 		/* Handle the case where the ASN.1 Path object specified
 		 * index and length values */
-		if (in_path->count < 0) {
-			if (file->size)
-				len = file->size;
-			else
-				len = 1024;
-			offset = 0;
-		}
-		else {
-			offset = in_path->index;
-			len = in_path->count;
-			/* Make sure we're within proper bounds */
-			if (offset >= file->size || offset + len > file->size) {
-				r = SC_ERROR_INVALID_ASN1_OBJECT;
+
+		if (file->ef_structure == SC_FILE_EF_LINEAR_VARIABLE) {
+
+			// in_path->index: record_no
+			// in_path->count: ignored!
+
+			if(file->record_length > 0) {
+				len = (file->record_length > MAX_FILE_SIZE)? MAX_FILE_SIZE:file->record_length + 5;
+			} else {
+				len = 0x2000 + 5;
```

Comment:
is this magic constant coming from somewhere specific or is it just arbitrary number? In any case, it should have some comment in the code so the people reading the code in couple of years do not have to ask the same question.